lib/pbench/server/api/resources/query_apis/metadata_index/index_mappings.py

```diff
@@ -56,23 +66,41 @@ class IndexMappings(Resource):
     }
     """
 
-    def __init__(self, logger: Logger):
-        self.logger = logger
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter(
+                    "index_key",
+                    ParamType.KEYWORD,
+                    required=True,
+                    keywords=list(RunIdBase.ES_INTERNAL_INDEX_NAMES.keys()),
+                    uri_parameter=True,
+                )
+            ),
+        )
 
-    def get(self, index_key: AnyStr) -> Dict[AnyStr, Any]:
+    def _get(self, index_key: AnyStr, request: Request) -> Dict[AnyStr, Any]:
         """
-        Return mapping properties of the document specified by the index name
-        in the URI. For example, user can get the run document properties by
-        making a GET request on index/mappings/search. Similarly other index
-        documents can be fetched by making a GET request on appropriate index names.
+        Return mapping properties of the document specified by the index key
+        in the URI (supported index keys are defined in RunIdBase.ES_INTERNAL_INDEX_NAMES).
+        For example, user can get the run document properties by making a GET
+        request on index/mappings/search. Similarly other index documents can be
+        fetched by making a GET request on appropriate index names.
 
         We fetch the mapping by querying the template database. If the
         template is not found in the database NOT_FOUND error will be raised.
         """
-        index = RunIdBase.ES_INTERNAL_INDEX_NAMES.get(index_key)
-        if not index:
-            self.logger.debug(f"Document index key is not valid {index_key!r}")
-            abort(HTTPStatus.NOT_FOUND, message="Mapping not found")
+        # Normalize and validate the index keys we got via URI string. These
+        # don't go through JSON schema validation, so we have
+        # to do it here.
+        try:
+            self.schema.validate(index_key)
+        except SchemaError as e:
+            abort(HTTPStatus.BAD_REQUEST, message=str(e))
```

Comment:
I'm confused..."JSON schema validation" validates only the _body_ of the request?

Reply:
`index_key` is a dictionary represented as `{"index_key": "<index_key_name_from_uri>"}` this is how the flask will pass the object to our `_get` method and since I am using the the same name for the `_get` parameter it becomes `index_key = {"index_key": "<index_key_name_from_uri>"}`.

And we validate this dictionary against the class schema that is defined as:
```
            Schema(
                Parameter(
                    "index_key",
                    ParamType.KEYWORD,
                    required=True,
                    keywords=list(RunIdBase.ES_INTERNAL_INDEX_NAMES.keys()),
                    uri_parameter=True,
                )
            )
```

Reply:
The answer to my question is, `ApiBase._dispatch()` calls `self.schema.validate()` but only for the body of the request, and only if the request is not a `GET`.

So, my follow-on question is, instead of calling `validate()` here, should we be enhancing `ApiBase._dispatch()` to handle that implicitly for us?  (@dbutenhof?)

Reply:
> So, my follow-on question is, instead of calling validate() here, should we be enhancing ApiBase._dispatch() to handle that implicitly for us? (@dbutenhof?)

I briefly considered wrapping query parameter validation into the framework when I added them for `GET /datasets/metadata` ... at the time that seemed like unnecessary complication. Now we've got payload body parameters, mapped Flask URI parameters, and query parameters ... and it'd be nice to have a coordinated standard mechanism to validate all of them. Hmm.

But, for the URI embedded parameters like this, we've apparently all forgotten that Nikhil did exactly that:

```
        try:
            if json_data:
                json_data.update(uri_parameters)
            else:
                json_data = uri_parameters
            new_data = self.schema.validate(json_data)
```

This updates the `Request.get_json()` dict with the `kwargs` query parameters supplied to `PUT`/`POST`, and validates them *all* at once. So simply adding `index_key` (and I agree with Webb that the name is perhaps a bit too "elastic-y") to the API's `Schema` should work just fine.

However ... due to the way Flask works, the parameters really aren't interchangeable; the Flask API mapping won't work if URI parameters are omitted, and I'm not sure I really like just lumping them all together.

I'm becoming tempted to enhance `_dispatch` to *separately* validate the `Request.get_json()` request body, the `kwargs` URI parameters, and the `Request.args` query parameters ... with separate `Schema` rather than lumping them all together.

But let's not blow up this PR, and I'll think about that later. For right now ... the `_dispatch` implementation is inconsistent with this subclass validation code, but it doesn't really conflict since the `_dispatch` validation will *ignore* the unknown `"index_key"` parameter rather than complaining...

Reply:
> for the URI embedded parameters like this, we've apparently all forgotten that Nikhil did exactly that

@dbutenhof, I don't think the code you referenced gets executed for a `GET` request, which is what I think we're talking about here.

> let's not blow up this PR, and I'll think about that later.

Fair enough.  (Should one of us create an issue to track this?)

Reply:
Yes, technically `GET` with URI parameters represents a bug in the prior PR because it misses the integrated validation step. Since I'd rather solve it a bit differently, it's OK not to fix that here. 😁 

I'll fork my response as an issue.